### DISCUSSION TITLE
how create the trafic.rar using traffic from TOMTOM or here

### DISCUSSION THREAD

**Discussion by Jofemago** [valhalla/valhalla#4256] [Q&A]
Hello . how are you? I'm trying to configure traffic in Valhalla for Colombia, and I have many doubts about traffic. I'm working with the TAR file, and I'm looking at TomTom's documentation. I can see that I can obtain flow information, but it's a bit confusing for me. It doesn't cover the whole city, just a part with a radius around a point. On the other hand, I can obtain traffic data using a polygon with HERE, but I find it quite confusing to define what should go inside that TAR file. Do you have any documentation that could provide me with more clarity? Thank you very much.

**Comment by itrack** [comment]
Hi @Jofemago , I also raised this question, maybe the answers will help you, I'm still looking for answers :D https://github.com/valhalla/valhalla/discussions/4412

**Comment by jthiard** [comment]
> but I find it quite confusing to define what should go inside that TAR file

Hi, I recently hacked with live traffic for a specific use cas of mine. What I can say is :

- there are no docs
- you need to understand C structs to get something working

The `traffic.tar` store traffic tiles (gph files, mirroring the routing tiles but with traffic data for each edge instead of routing data).

Each tile has:
- a header (https://github.com/valhalla/valhalla/blob/3.5.0/valhalla/baldr/traffictile.h#L184-L191)
- a traffic speed struct for each routing edge (https://github.com/valhalla/valhalla/blob/3.5.0/valhalla/baldr/traffictile.h#L52-L64)

As you can see the speed struct allow you to set multiple live speeds, by cutting the edge in max 3 part (2 breakpoints). If you want to use the same speed along the whole edge, you can set the first breakpoint value to 255 (max encoded length).

Correspondance between those tiles and the routing tiles is important. So you need to recreate the traffic tar each time you update your routing data.

**Create the traffic.tar**

That part is easy. the script `valhalla_build_extract` has an option `--with-traffic` that will generate the tar skeleton (tiles with their headers, and reserve zero bytes for the edges (https://github.com/valhalla/valhalla/blob/3.5.0/scripts/valhalla_build_extract#L317-L361).

**Update edges with live speeds**

This is the hard part, you are basically on your own. The goal is to open the tar, iterate over edges in each tile, and update speed structs depending of your data.
While it could theoretically be done with whatever language that support dealing with tar files and C structs (eg. python), I found it easier to do it in C++ to benefit from the `GraphReader` and `GraphTile` interfaces when iterating over edges.
You can find an example of how to open the tar an iterate over edges in the tests (https://github.com/valhalla/valhalla/blob/3.5.0/test/test.cc#L514-L556)

**Reply by Jofemago** [reply]
thanks bro, I am going to check this but I do not know when xd, but it is a great advance, thank you!

**Comment by kevinkreiser** [comment]
yeah we provide the structs etc in the Valhalla library but we don't provide a tool to directly create traffic tiles with real data simply because there is no clean way that I can think of to support all use cases. like we don't know where your traffic data comes from or what it looks like. the best we could do is write code with an api for updating an existing tar but I know from experience this is certainly not a scalable solution for at least a couple different configurations of traffic providers. Because of this we give you the library, the structs and the router to use the structs but don't prescribe how do to it because the prescription wouldn't make sense for all setups. I've written this in other issues before but should probably pin it somewhere since it keeps coming up.

**Reply by culebron-ala** [reply]
I was searching through the docs, expecting something like OSRM has: I must give it speeds on every edge/segment, and if I don't, it takes the speed from the rules (OSM tags). And it's my duty to keep and maintain the data -- the router doesn't try to be smart. I see it becomes harder if you try to set speeds for parts of the edge length -- but I was prepared to see it just like `edge_id,speed`, and to live with it. :)

**Reply by nilsnolde** [reply]
Our way of producing speed segments is much more detailed, so start and end are more proper. You don't HAVE TO use the segments, you can just use one speed for the entire edge by not setting breakpoints and simply the first speed.

**Reply by kinkard** [reply]
>but I was prepared to see it just like edge_id,speed, and to live with it.

This is exactly how live-traffic works in Valhalla - find a proper offset in a `traffic.tar` using `edge_id` and then assign speed to it.
The only nuance is that you will need to either include https://github.com/valhalla/valhalla/blob/master/valhalla/baldr/traffictile.h into your project to work with memory-mapped tar directly or copy-paste all important pieces in your language of choise to do the same.

Also worth noting that edge IDs are not stable over tiles builds (well, same config and same OSM PBF basemap should produce same tiles, but no guarantees if config or basemap changes) and that map-matching (that you need to use to get edge IDs) still [has several areas for improvement](https://github.com/orgs/valhalla/projects/2).
